storage/mempool.go

```diff
@@ -0,0 +1,817 @@
+// exist in the state tree. In that case this function will return a storage.NotFoundError
+func (s *Storage) getPendingState(stateID uint32) (
+	nonce *models.Uint256,
+	balance *models.Uint256,
+	err error,
+) {
+	// TODO: not confident this does the right thing, this wants some tests
+
+	key := pendingStateKey(stateID)
+	value, err := s.rawLookup(key)
+
+	if err == nil {
+		decodedNonce, decodedBalance := decodePendingState(value)
+		return &decodedNonce, &decodedBalance, nil
+	}
+
+	if errors.Is(err, badger.ErrKeyNotFound) {
+		state, innerErr := s.StateTree.Leaf(stateID)
+		if innerErr != nil {
+			return nil, nil, innerErr
+		}
+
+		nonce = &state.UserState.Nonce
+		balance = &state.UserState.Balance
+		return nonce, balance, nil
+	}
+
+	return nil, nil, err
+}
+
+func (s *Storage) setPendingState(stateID uint32, nonce, balance models.Uint256) error {
+	key := pendingStateKey(stateID)
+	value := encodePendingState(nonce, balance)
+	return s.rawSet(key, value)
+}
+
+func (s *Storage) addToPendingBalance(stateID uint32, amount *models.Uint256) error {
+	pendingNonce, pendingBalance, err := s.getPendingState(stateID)
+	if err != nil {
+		return err
+	}
+
+	newBalance := pendingBalance.Add(amount)
+
+	return s.setPendingState(stateID, *pendingNonce, *newBalance)
+}
+
+func (s *Storage) GetPendingNonce(stateID uint32) (*models.Uint256, error) {
+	pendingNonce, _, err := s.getPendingState(stateID)
+	return pendingNonce, err
+}
+
+func (s *Storage) GetPendingBalance(stateID uint32) (*models.Uint256, error) {
+	_, pendingBalance, err := s.getPendingState(stateID)
+	return pendingBalance, err
+}
+
+func (s *Storage) GetPendingUserState(stateID uint32) (*models.UserState, error) {
+	leaf, err := s.StateTree.Leaf(stateID)
+	if err != nil {
+		return nil, err
+	}
+
+	pendingNonce, pendingBalance, err := s.getPendingState(stateID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &models.UserState{
+		PubKeyID: leaf.UserState.PubKeyID,
+		TokenID:  leaf.UserState.TokenID,
+		Balance:  *pendingBalance,
+		Nonce:    *pendingNonce,
+	}, nil
+}
+
+// TODO: this assumes the sender & receiver stateIDs are in the state tree,
+//       make sure all callers check for this? Maybe return a nice error?
+func (s *Storage) AddMempoolTx(tx models.GenericTransaction) error {
+	// TODO: should we check that this txn does not already exist as a batchedTx?
+	// TODO: should this accept a stored.PendingTx, so we do not accidentally accept
+	//       a tx which already has a CommitmentSlot?
+
+	// The caller should have opened a txn but doing it here just in case.
+
+	// The txn gives badger enough information to lock out conflicting API handlers
+	// for us. storage/database.go is even nice enough to automatically retry the
+	// fn for us in case of ErrConflict, so `unsafeAddMempoolTx` had better be
+	// idempotent.
+
+	return s.ExecuteInReadWriteTransaction(func(txStorage *Storage) error {
```

Comment:
instead of this wrap we should probably error out if you're *not* in a transaction